### ISSUE TITLE
global shared variable

### ISSUE THREAD

**Issue by shockjiang** [halide/Halide#2888]
I'm trying to define a global shared variable, which counts a non-zero elements in a input, like this:

counter = 0
N += select(input[x,y] > 10, 1, 0)

however, this is very hard in halide, is there any global shared variable that targets this goal?

**Comment by MDBrothers** [comment]
Couldn't you at least count along rows in parallel and then do a serial reduction on the resulting 1D vector?
```

Halide::RDom strip(0, 16);

Halide::Func binarized("binarized");
binarized(x,y) = Halide::select( input(x, y) > 10, 1, 0);
Halide::Func binarizedWithBC = Halide::boundarcondition::constant_exterior(binarized, 0, 0, imageWidth, 0, imageHeight);

Halide::Func count1, count2, count3.;
count1(x, y) = Halide::sum(binarizedWithBC (strip.x + 16*x, y));
count1.parallel(y);

count2(x, y) = Halide::sum(count1(strip.x + 16*x, y));
count2.parallel(y);

count3(x, y) = Halide::sum(count2(strip.x + 16*x, y));
count3.parallel(y);

Halide::RDom finalR(0, imageWidth / 16/ 16/ 16+ 1, 0, imageHeight);

Halide::Expr finalCount = Halide::sum( count3(finalR.x, finalR.y));

```

Crude probably but fast enough.

**Comment by ashishUthama** [comment]
The parallel histogram section in the [Scheduling FAQ](https://github.com/halide/Halide/wiki/Scheduling-FAQ) shows a way similar to what @MDBrothers mentioned - you could adapt that too (Since you essentially need 'one bucket')).

**Comment by MDBrothers** [comment]
I agree with @ashishUthama . The halide scheduling example as well as the associative reduction tutorial http://halide-lang.org/tutorials/tutorial_lesson_18_parallel_associative_reductions.html I think are related to your problem. I'm always pleasantly surprised when almost anything reasonable works, scheduling wise.

**Comment by abadams** [comment]
**Automated triage note:** this issue was flagged by Claude as likely safe to close, for the reason below. The reasoning was checked against the current source, and a second pass tried and failed to refute it, but no human verified this specific text. **If the reason is wrong, please reopen the issue and say so** — that is the expected outcome for some fraction of these, and no offence taken.

**Reason (CLOSE_WONTFIX):** Asks for a mutable global accumulator. Halide has no such construct by design; the supported way to express a global count is a reduction over an RDom (optionally split into a parallel associative reduction), which two commenters explained with working examples. Nothing actionable in the code.

**Evidence:** tutorial/lesson_18_parallel_associative_reductions.cpp and the Scheduling FAQ wiki's parallel-histogram section, both still current; no global-variable construct exists in src/

Closing: Halide has no mutable global variable by design — express this as a reduction over an RDom (sum(select(input(r.x, r.y) > 10, 1, 0))), and see tutorial/lesson_18_parallel_associative_reductions.cpp if you need it parallelised.
